[Bandov] iP

Bandov opened the pull request:
### **BobBot**

> The More Time You Spend Awake The More Time You Spend Asleep. - Jaden Smith

Bob cooks, cleans and keeps your quintessential information for you. It's

- text-based
- fast-paced
- ~~easy~~ **super easy** to use

To get bob in your life you need to

1. download him from [here](https://github.com/Bandov/ip) 
2. run the installer
3. tell him your tasks
4. free money 😉

All for the cost of **$0**!

_Things bob can do!_

- [ ] cook
- [ ] clean
- [x] Manage tasks
- [x] Manage deadlines
- [x] Manage events

If you are a programmer come join me! Here is the main code for you to run
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Bandov committed:
Level 0

Bandov committed:
Level 1

Bandov committed:
Level 2

Bandov committed:
Level 3

Bandov committed:
Level 4

Bandov committed:
fixed double lining

Bandov committed:
Added error catching functionality

Bandov committed:
deleting of tasks from task list

Bandov committed:
Add the autorunning of tests

Bandov committed:
correct testing methodology

Bandov committed:
added saving

Bandov committed:
added time functionality

Bandov committed:
OOP

Bandov committed:
added to package duke

Bandov committed:
Merge commit 'a6f73244a20539e20c2189973ebf597df34ef631' into A-Packages

Bandov committed:
Added gradle testing and functionality

Bandov committed:
no message

Bandov committed:
Added more JUnit tests

Bandov committed:
Jar file creation

Bandov committed:
gradle checkstyle plus javadocs

Bandov committed:
coding standard

Bandov committed:
added find capability

Bandov committed:
changed indentation to follow checkstyle

Bandov committed:
fix wrong branching

Bandov committed:
Merge commit 'e7485a33153f814dddfc4cdd82dfef9d3e4ab8a3' into branch-Level-8

* commit 'e7485a33153f814dddfc4cdd82dfef9d3e4ab8a3':
  fix wrong branching

Bandov committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  fix wrong branching

# Conflicts:
#	src/main/java/Duke.java

Bandov committed:
Updated to fit checkstyle requirements

Bandov committed:
implemented functional user interface

Bandov committed:
fixed checkstyle issues

Bandov committed:
added missing javadocs

Bandov committed:
fixed the no storage file issue

Bandov committed:
made duke string path final and implemented varargs

Bandov committed:
created a jar file to be used

Bandov committed:
Add assertions to improve code robustness

Incorporate assertions throughout the codebase to ensure key assumptions about variable states and method outcomes hold true. This addition aims to catch and identify potential bugs early in the development cycle, enhancing the overall reliability and maintainability of the project. Assertions have been added primarily in critical sections where unexpected values could lead to logic errors or unexpected behavior.

- Add assertions in method argument checks to prevent invalid inputs.
- Ensure data integrity by asserting expected conditions before and after data manipulation.
- Validate state consistency within objects post-initialization and during key operations.

This update does not affect the external API or performance under normal operation conditions but significantly improves code quality and debuggability.

Bandov committed:
Merge pull request #1 from Bandov/branch-A-Assertions

Add assertions to improve code robustness

Bandov committed:
Comprehensive code quality improvement across the project

- Standardized code formatting and naming conventions for improved consistency.
- Modularized large components and optimized data structures for better maintainability and performance.
- Enhanced encapsulation and applied design patterns where applicable for increased robustness and flexibility.
- Implemented dependency injection to decouple components, facilitating easier testing and modification.
- Refined error handling and logging mechanisms to improve application reliability and debuggability.
- Updated and expanded documentation, including Javadoc and in-line comments, for clearer understanding and maintainability.
- Increased test coverage with unit and integration tests to ensure code reliability and facilitate safe refactoring.
- Addressed security concerns and vulnerabilities to protect against potential threats.

This refactor aims to enhance the overall quality of the project, focusing on readability, maintainability, performance, and security, ensuring a solid foundation for future development and scalability.

Bandov committed:
Merge pull request #2 from Bandov/branch-A-CodeQuality

Comprehensive code quality improvement across the project

Bandov committed:
Add GitHub Actions for Continuous Integration

Implemented GitHub Actions to automate the build and testing process for the project. This includes setting up workflows for running unit tests and linting on every push and pull request to the main branch. The addition of CI ensures that code changes are automatically verified, maintaining code quality and stability across contributions.

- Added `.github/workflows/gradle.yml` defining the CI pipeline.

Bandov committed:
Create gradle.yml

Bandov committed:
Merge pull request #3 from Bandov/Bandov-patch-1

Create gradle.yml

Bandov committed:
Fixed a mistake in pushing gradle.yml

Bandov committed:
Merge branch 'master' of https://github.com/Bandov/ip

* 'master' of https://github.com/Bandov/ip:
  Create gradle.yml

Bandov committed:
feat: Implement tagging feature for tasks

This commit introduces the ability for users to add tags to tasks using the format 'tag [task number] /tag [tag content]'. Tags are displayed alongside task descriptions and are prefixed with a '#'. The handleTag method has been updated to parse the command and append the tag to the specified task.

GiselleTan commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,88 @@
+     */
+    @Override
+    public String toString() {
+        return "[D]" + (isDone ? "[X] " : "[ ] ") + super.description + " (by: " + by + ")"
```
I noticed that you're using the (isDone ? "[X] " : "[ ] ") in the toString() and toFileString() functions as well as in other functions in other classes. Perhaps create a function such as printCurrentStatus() to accomplish this?

GiselleTan commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,109 @@
+     * @return An Event object, or null if the string format is invalid.
+     */
+    public static Event fromFileString(String str) {
+        String[] parts = str.split(" \\| ");
```
Naming it as "parts" can be confusing. Perhaps naming it as eventDetails would make it clearer what the String being split is.

GiselleTan commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,109 @@
+     */
+    public static Event fromFileString(String str) {
+        String[] parts = str.split(" \\| ");
+        if (!parts[0].equals("E")) {
```
Putting parts[0] in a variable might make it clearer what it is referring to and more readable.

GiselleTan reviewed (commented):
Overall, your comments helped with understanding your code a lot and the detail of the comments must have taken a lot of time!

1rbg commented on src/main/java/duke/Ui.java:
Some minor inconsistency with the naming here, the first method is `void printWithLines()` which is fine and we expect it to print something behind the abstraction. The second method is `void showLoadingError()`, which is also fine since it will serve as a UI abstraction that shows the error on the UI. However, the last method is `String showWelcome()`, which also starts with show, which would indicate serving as a UI abstraction that shows the welcome, however it just returns a String. Perhaps it could be renamed `getWelcomeString()`, or the implementation could directly use `printWithLines()`.

1rbg commented on src/main/java/duke/Task.java:
I like how you have named `toFileString()` and `fromFileString()` here, it is clear what it is used for and how it can be used.

1rbg commented on src/main/java/duke/Storage.java:
I like that your imports are in alphabetical order and not using wildcard imports, this is consistent and follows the coding standard!

1rbg reviewed (commented):
Overall very well done! In terms of coding standard, I think you have done very well. Keep it up!

Bandov committed:
Enhance GUI layout and accessibility

Bandov committed:
Add comprehensive unit tests for Event class functionality

Bandov committed:
Add README documentation for the tagging feature

This commit introduces a README file that details the tagging feature within the application. It covers the creation and usage of tags for categorizing tasks, including methods for associating, filtering, and removing tags from tasks. This documentation aims to provide users and developers with a clear understanding of how tags can be used to enhance task management and organization.

Bandov committed:
Updated to fix the image

Bandov committed:
Fixed image issues

Bandov committed:
Fix Image issues

Bandov committed:
fix image issues

Bandov committed:
Cleanup of readme file

Bandov committed:
fixed image issues

Bandov committed:
fixed the parser issues

Bandov committed:
fixed the UG and bug fixes

Bandov committed:
updated image

Bandov committed:
bug fix
